number.multiple() enhancement - Issue #916

santanu-biswas opened the pull request:
‘number.multiple()’ enhanced for:
1. Accept any Number (integer or float) as argument.
2. Accept references as argument.
3. For any non-number or zero multiple base, validate now returns error
instead of throwing it.

santanu-biswas commented:
Rebase done

Marsup commented:
You need to check error messages as well, false is not enough (eg. https://github.com/hapijs/joi/blob/master/test/number.js#L548). That's a new policy I intend to apply to all the tests but didn't have time to do.

Marsup commented on lib/number.js:
```diff
@@ -53,12 +58,31 @@ internals.Number = class extends Any {
+            }
 
-        return this._test('multiple', base, function (value, state, options) {
+            const mult = Math.pow(10, Math.max(internals.precision(value), internals.precision(divisor)));
```
Which case it this trying to prevent ?

santanu-biswas replied:
@Marsup 
It is to allow number.multiple() to allow decimal numbers. However % operator cannot handle decimal numbers correctly. So the idea is to change the decimal numbers to integers before using % operator.

Example:

- To check if 4.523 is a multiple of 1.3, it checks if 4523 (4.523 * 10^3) is a multiple of 1300 (1.3 * 10^3).
- To check if 4.2 is a multiple of 1.05, it checks if 420 (4.2 * 10^2) is a multiple of 105 (1.05 * 10^2)

Added internals.precision(num) to find the number of decimal places in a any number.

Marsup replied:
Is it a real world problem for existing use cases though ? We don't want accurate modulo, we just want 0 when it's a multiple.

santanu-biswas replied:
Yes @Marsup it is a real life problem for existing use cases in financial application involving exchange rates that can have 2 to 5 decimal accuracy.

number.multiple() fails to validate if an entered currency value is exact multiple of an amount in another currency.

**Example:**
6711 % 67.11 results in 5.684341886080802e-14
3230 % 32.3 results in 2.8421709430404007e-13

_Although in both the above cases it should have been 0 but modulo operator is broken for decimal/float values. (Check [JSFiddle](https://jsfiddle.net/santanubiswas/fL1zeadp/))_

Similar real world use cases also exist in stock trading application.

santanu-biswas commented:
@Marsup 
I have added the required error message checks for tests that return false.

Marsup commented:
I'm now wondering whether it should be an extension or not, it's still unsafe and we'd be better off using something like big.js/decimal.js or whatever module works.

santanu-biswas commented:
Initially I did think about adding a module float.js, however, then I realized that number.js will have to be modified to handle only integers. Otherwise, there would be an ambiguity between Joi.number() and Joi.float(). But then if this change is done then it would break current usage code wherever Joi.number() is being used to validate integers.

Thats why I felt that for now doing it in number.js will be better and then upgrade by adding float.js to clearly distinguish between integer and float validation.

Marsup commented:
You misunderstood me, I was talking about modules, not files. I don't see any point in having another type since javascript doesn't make this difference either.

santanu-biswas commented:
@Marsup 
OK so what do you suggest now?
What makes you feel that it is unsafe?

Marsup commented:
Sorry not unsafe, unreliable. However unlikely, given enough digits, you'll overflow the maximum capacity of javascript numbers with the multiplication. I'm still looking whether we can achieve the same thing without increasing original values but I can't find anything.

santanu-biswas commented:
@Marsup I understand your point now.
So, is it alright if I change it to simple modulus like before and keep the other enhancements:
1. Accept references as argument.
2. For any non-number or zero multiple base, validate now returns error
instead of throwing it.

santanu-biswas commented:
@Marsup 
Please have a look at [_**decimal.js**_](https://www.npmjs.com/package/decimal.js). It has a robust & reliable _mod_ function. 

(A lighter version of it is also available at [_**decimal.js-light**_](https://github.com/MikeMcl/decimal.js-light/)).

I tested the _mod_ function with fairly large numbers (See [JSFiddle](https://jsfiddle.net/santanubiswas/khgns539/)) and it works perfectly alright.

If you agree, it can be used.

Marsup commented:
I already mentioned that library a few comments ago, it's also why I said I might have it as an extension, since this is exactly why it was made.

santanu-biswas commented:
@Marsup 
Added big.js and made the required changes in modulo function and completed all tests.

Marsup commented:
When I said an extension, I meant a separate module (see [Joi.extend()](https://github.com/hapijs/joi/blob/master/API.md#extendextension)), sorry if that was unclear. I'm not adding any dependencies to the core unless they serve a common interest.
What can be done in the core is accepting floats and references though, but documenting that it has the same flaws as javascript itself.

santanu-biswas commented:
@Marsup Okay... I understand. Removed _big.js_ but kept float acceptance and refs.
Also added Notes to documentation as suggested by you. Hope I got it alright now.

Marsup commented on lib/number.js:
```diff
@@ -53,12 +53,29 @@ internals.Number = class extends Any {
 
     multiple(base) {
 
-        Hoek.assert(Hoek.isInteger(base), 'multiple must be an integer');
```
Don't remove those, just allow floats.

santanu-biswas replied:
If you don't remove this then it throws error instead of gracefully validating (thats how it happens even now). It starts throwing exception for non-integer number, so what is the point in allowing float numbers?

Marsup replied:
And that's intended, this is to prevent developers mistakes.

santanu-biswas replied:
So that means:

- The developer should first validate if the multiple is zero (the developer is using Joi for validation and there is no way to validate that in Joi).
- Allowing floats has no meaning, because it will throw exception anyways.

Am I right?
So in effect we are just making the changes for allowing refs.?

Marsup replied:
The developer is using joi to validate external inputs, not to validate its own inputs, there are synchronous throws for that yes.
How would allowing floats crash anything ?

Marsup commented on lib/number.js:
```diff
@@ -53,12 +53,29 @@ internals.Number = class extends Any {
+            let divisor;
+            if (isRef) {
+                divisor = base(state.parent, options);
+                if (typeof divisor !== 'number') {
```
<= 0 would also be an error.

Marsup replied:
Also check for NaN

Marsup commented on lib/number.js:
```diff
@@ -53,12 +53,29 @@ internals.Number = class extends Any {
+            }
+            else {
+                divisor = base;
+                if ((typeof divisor !== 'number') || (divisor === 0)) {
```
This if should be removed since conditions would be checked upfront.

santanu-biswas replied:
Are you asking for removing 

`if (divisor === 0) {`

If this is removed then It starts throwing error if the argument is zero.  The idea is to return if the number.multiple() test passes or not instead of throwing error. The idea is to validate gracefully instead of throwing errors.

Marsup commented on lib/number.js:
```diff
@@ -53,12 +53,29 @@ internals.Number = class extends Any {
+        const isRef = Ref.isRef(base);
 
-        return this._test('multiple', base, function (value, state, options) {
+        return this._test('multiple', base, (value, state, options) => {
```
Do not use => here.

Marsup commented on lib/number.js:
```diff
@@ -53,12 +53,22 @@ internals.Number = class extends Any {
+        const isRef = Ref.isRef(base);
+
+        if (!isRef) {
+            Hoek.assert(Hoek.isInteger(base), 'multiple must be an integer');
```
So why aren't you allowing floats here ?

santanu-biswas replied:
@Marsup Allowed floats now.

santanu-biswas commented:
@Marsup Have allowed all numbers (Integers & floats). Is there anything else that I have missed?

Marsup commented:
I don't think so but I'll review more in depth tonight.

Marsup commented:
Merged as 3a32b92. Cleaned it up a bit before merging, thanks for your contribution.
